include/tvm/relay/vm/vm.h

```diff
@@ -131,65 +165,62 @@ void VMFunctionPrint(const VMFunction& vm_func);
  */
 struct VMFrame {
     size_t pc;
-    size_t bp;
-    size_t sp;
     size_t func_index;
     size_t args;
     const Instruction* code;
 
-    VMFrame(size_t pc, size_t bp, size_t sp, size_t func_index, size_t args, const Instruction* code)
-      : pc(pc), bp(bp), sp(sp), func_index(func_index), args(args), code(code) {}
+    // Points to the offset of this function's register file start inside VM's register stack
+    size_t register_stack_start;
+
+    // map from virtual register to memory slot
+    std::unordered_map<VirtualRegisterNum, SlotNum> register_file_map;
+
+    VMFrame(size_t pc, size_t func_index, size_t args, const Instruction* code, size_t register_stack_start,
+            std::unordered_map<size_t, size_t> register_file_map)
+      : pc(pc), func_index(func_index), args(args), code(code), register_stack_start(register_stack_start),
+        register_file_map(register_file_map) {}
 };
 
 struct VirtualMachine {
     // TODO(@jroesch):
     std::vector<PackedFunc> packed_funcs;
     std::vector<VMFunction> functions;
     std::vector<VMFrame> frames;
-    std::vector<Object> stack;
     std::vector<Object> constants;
 
     // Frame State
     size_t func_index;
     const Instruction* code;
     size_t pc;
-    size_t bp;
+    // Top position of register stack
+    size_t top_stack;
+    // Special register to save function call return value
+    Object return_register;
```

Comment:
Put `return_register` in the `VMFrame` so it's possible that we can run `VMFunction` in parallel.

Reply:
When `Ret` executed, the frame is poped out. So result will be gone before next instruction gets executed.